### PR TITLE
[Merged by Bors] - hashing: switch from sha256 to blake3

### LINKED ISSUE #1275: Unify hashing algorithms
Our code uses at least 5 different hashing algorithms (`Sha256`, `Sha3`, `Keccak256`, `Shake128`, `FNV`). We should decide on a hashing algo and replace all the others in the codebase.

Let's use https://github.com/spacemeshos/sha256-simd ?

### PR DESCRIPTION
**Pull request by countvonzero**
## Motivation

Closes #1275

## Changes
use blake3 instead of sha256 for hashing.

### CODE REVIEW

**Comment by countvonzero** [comment]
bors try

**Comment by countvonzero** [comment]
bors try

**Review by dshulyak — APPROVED** [review]

File: common/types/layer.go [hunk 1]
```diff
@@ -14,6 +14,7 @@ const (
 	LayerIDSize = 4
 
 	// BootstrapBeacon is the hex value of the beacon used during genesis.
+	// TODO: make this a config.
 	BootstrapBeacon = "0x333c04dd151a2a6831c039cb9a651df29198be8a04e16ce861d4b6a34a11c954" // sha256("bootstrap")
```
Comment (poszu) [inline common/types/layer.go:18]:
```suggestion
	BootstrapBeacon = "0x333c04dd151a2a6831c039cb9a651df29198be8a04e16ce861d4b6a34a11c954" // blake3("bootstrap")
```

File: hare/preroundtracker.go [hunk 2]
```diff
@@ -49,21 +49,21 @@ func (pre *preRoundTracker) OnPreRound(ctx context.Context, msg *Msg) {
...
-	shaUint32 := binary.LittleEndian.Uint32(sha[:4])
+	vrfHash := hash.Sum(msg.Eligibility.Proof)
+	vrfHashVal := binary.LittleEndian.Uint32(vrfHash[:4])
 	logger.With().Debug("received preround message",
 		nodeID,
 		log.String("sender_id", nodeID.ShortString()),
 		log.Int("num_values", len(msg.InnerMsg.Values)),
-		log.String("vrf_value", fmt.Sprintf("%x", shaUint32)))
+		log.Uint32("vrf_value", vrfHashVal))
 	// TODO: make sure we don't need a mutex here
```
Comment (poszu) [inline hare/preroundtracker.go]: That's a serious TODO. I created an issue to track it. Would you mind sticking a link in the comment while you are at it? https://github.com/spacemeshos/go-spacemesh/issues/4007

**Review by poszu — APPROVED** [review]

File: hare/preroundtracker.go (hunk 2 above)
Comment (countvonzero) [reply hare/preroundtracker.go]: we dont need a mutex here. if you see algorithm.go, you will see messages are handled via a single channel

File: common/types/layer.go (hunk 1 above)
Comment (countvonzero) [reply common/types/layer.go:18]: but thats not true. that value is generated via sha256. in any case, that constant needs to go to a config

File: hare/preroundtracker.go (hunk 2 above)
Comment (countvonzero) [reply hare/preroundtracker.go]: removed this comment instead and commented on issue #4007

**Comment by countvonzero** [comment]
bors merge

**Review by fasmat — APPROVED** [review]

File: common/types/layer.go (hunk 1 above)
Comment (poszu) [reply common/types/layer.go:18]: OK, I didn't know. That's a bit confusing. Why is it the only thing that uses sha256?

Comment (countvonzero) [reply common/types/layer.go:18]: it was generated when everything else uses sha256. now we are changing to blake3. i just don't see a point of regenerating this with blake3 when it really should live in the config. hence the TODO.

**Comment by countvonzero** [comment]
bors merge

**Comment by poszu** [comment]
bors merge

**Comment by countvonzero** [comment]
bors merge

**Comment by countvonzero** [comment]
bors merge

**Pull request closed without merging**